ggml/src/ggml-cpu/CMakeLists.txt

```diff
@@ -151,30 +151,44 @@ function(ggml_add_cpu_backend_variant_impl tag_name)
                 endif ()
             endif ()
         else()
-            check_cxx_compiler_flag(-mfp16-format=ieee COMPILER_SUPPORTS_FP16_FORMAT_I3E)
-            if (NOT "${COMPILER_SUPPORTS_FP16_FORMAT_I3E}" STREQUAL "")
-                list(APPEND ARCH_FLAGS -mfp16-format=ieee)
-            endif()
-            if (${CMAKE_SYSTEM_PROCESSOR} MATCHES "armv6")
-                # Raspberry Pi 1, Zero
-                list(APPEND ARCH_FLAGS -mfpu=neon-fp-armv8 -mno-unaligned-access)
-            endif()
-            if (${CMAKE_SYSTEM_PROCESSOR} MATCHES "armv7")
-                if ("${CMAKE_SYSTEM_NAME}" STREQUAL "Android")
-                    # Android armeabi-v7a
-                    list(APPEND ARCH_FLAGS -mfpu=neon-vfpv4 -mno-unaligned-access -funsafe-math-optimizations)
-                else()
-                    # Raspberry Pi 2
-                    list(APPEND ARCH_FLAGS -mfpu=neon-fp-armv8 -mno-unaligned-access -funsafe-math-optimizations)
+            if (GGML_NATIVE)
+                set(USER_PROVIDED_MARCH FALSE)
+                foreach(flag_var IN ITEMS CMAKE_C_FLAGS CMAKE_CXX_FLAGS CMAKE_REQUIRED_FLAGS)
+                    if ("${${flag_var}}" MATCHES "-march=[a-zA-Z0-9+._-]+")
+                        set(USER_PROVIDED_MARCH TRUE)
+                        break()
+                    endif()
+                endforeach()
+
+                if (NOT USER_PROVIDED_MARCH)
+                    list(APPEND ARCH_FLAGS "-march=native")
+                endif()
+            else()
```

Comment:
A bit too drastic maybe but I guess the other cases are only relevant when cross-compiling.

Reply:
Yes, the other code in fact seems to be doing the same thing that `-march=native` would do. `GGML_NATIVE` disabled should generate a consistent build depending on the flags specified during compilation, which is not the case at the moment.

This needs to be completely revamped, and as it is, this PR is just adding to the mess that will need to be cleaned up later.

Reply:
I can help with revamping but I need some clarification first :)

Today, building on generic ARM gives very poor performance because the build system completely ignores the `GGML_NATIVE` directive, so I just aligned the current code with the current description of `GGML_NATIVE` found in `ggml/CMakeLists.txt` :
```
option(GGML_NATIVE "ggml: enable -march=native flag" ${GGML_NATIVE_DEFAULT})
```
I completely agree that it's not the best way to get performance but it's better than nothing and it already fixes many modern setups.

So, do we want to relax the definition of `GGML_NATIVE` and allow to use, for example, `-mcpu=native` on ARM which would be much better for performance?

The old code was clearly aimed at small devices, like android and raspberry and also used `CMAKE_SYSTEM_PROCESSOR` so for me it wasn't used as a way to fix `-march=native` at all but rather as a way to find acceptable flags when cross-compiling and in this case you really don't want `GGML_NATIVE` (hence the move to `else`).

Maybe @ggerganov has some memories to share about that ?

Reply:
If I understand correctly, with gcc/clang it is enough to set the correct architecture flags with `-march`, and `-match=native` should work in the same way as x86. The exception is likely to be MSVC once again, because it does not set the preprocessor definitions of the enabled ARM features. In which case, we may consider just dropping support for MSVC with ARM entirely, because it is a constant source of problems, doesn't work with the inline asm kernels, and doesn't really add anything beyond clang or possibly MINGW.

I believe this should work:
- Set `-march=native` if `GGML_NATIVE` is enabled
- Add a parameter `GGML_CPU_ARCH` to the build to set the architecture, so that if `GGML_NATIVE` is disabled and this parameter is provided, then `-march=${GGML_CPU_ARCH}` is used.

Reply:
On ARM, to build for local use (i.e. using `GGML_NATIVE`), `-mcpu=native` alone should be the best option as far as I know. The `-march=native` will often miss some opportunities. The `-mtune=native` will optimize for the current microarchitecture (so still not fully optimized for the cpu).

So I think redefining `GGML_NATIVE` to something like "Try to optimize builds for the current cpu" and using `-march=native` on `x86_64` and `-mcpu=native` on `ARM` would already be much simpler and an improvement.

Reply:
If this sounds good to you, I can adapt this PR in this direction so we can see how it works in practice.

Reply:
Yes, sounds good.